cardinal/router/router.go

```diff
@@ -0,0 +1,238 @@
+			Errs: fmt.Errorf("unable to find persona tag associated with the EVM address %q: %w", req.Sender, err).
+				Error(),
+			EvmTxHash: req.EvmTxHash,
+			Code:      CodeUnauthorized,
+		}, nil
+	}
+
+	// since we are injecting the msgValue directly, all we need is the persona tag in the signed payload.
+	// the sig checking happens in the server's Handler, not in ecs.Engine.
+	sig := &sign.Transaction{PersonaTag: personaTag}
+	r.provider.AddEVMTransaction(msgType.ID(), msgValue, sig, req.EvmTxHash)
+
+	// wait for the next tick so the msgValue gets processed
+	success := r.provider.WaitForNextTick()
+	if !success {
+		return &routerv1.SendMessageResponse{
+			EvmTxHash: req.EvmTxHash,
+			Code:      CodeServerUnresponsive,
+		}, nil
+	}
+
+	// check for the msgValue receipt.
+	result, errs, evmTxHash, exists := r.provider.ConsumeEVMMsgResult(req.EvmTxHash)
+	if !exists {
+		return &routerv1.SendMessageResponse{
+			EvmTxHash: req.EvmTxHash,
+			Code:      CodeNoResult,
+		}, nil
+	}
+
+	// we got a receipt, so lets clean it up and return it.
+	var errStr string
+	code := CodeSuccess
+	if retErr := errors.Join(errs...); retErr != nil {
+		code = CodeTxFailed
+		errStr = retErr.Error()
+	}
+	return &routerv1.SendMessageResponse{
+		Errs:      errStr,
+		Result:    result,
+		EvmTxHash: evmTxHash,
+		Code:      code,
+	}, nil
+}
+
+// QueryShard is the server impl that answers query requests from the base shard client.
+func (r *routerImpl) QueryShard(_ context.Context, req *routerv1.QueryShardRequest) (
+	*routerv1.QueryShardResponse, error,
+) {
+	zerolog.Logger.Debug().Msgf("get request for %q", req.Resource)
+	reply, err := r.provider.HandleEVMQuery(req.Resource, req.Request)
+	if err != nil {
+		zerolog.Logger.Error().Err(err).Msg("failed to handle query")
+		return nil, err
+	}
+	zerolog.Logger.Debug().Msgf("sending back reply: %v", reply)
+	return &routerv1.QueryShardResponse{Response: reply}, nil
+}
+
+func (r *routerImpl) Submit(
+	ctx context.Context,
+	processedTxs txpool.TxMap,
+	namespace string,
+	epoch,
+	unixTimestamp uint64,
+) error {
+	messageIDtoTxs := make(map[uint64]*shard.Transactions)
+	for msgID, txs := range processedTxs {
+		protoTxs := make([]*shard.Transaction, 0, len(txs))
+		for _, txData := range txs {
+			protoTxs = append(protoTxs, transactionToProto(txData.Tx))
+		}
+		messageIDtoTxs[uint64(msgID)] = &shard.Transactions{Txs: protoTxs}
+	}
+	req := shard.SubmitTransactionsRequest{
+		Epoch:         epoch,
+		UnixTimestamp: unixTimestamp,
+		Namespace:     namespace,
+		Transactions:  messageIDtoTxs,
+	}
+	_, err := r.ShardSequencer.Submit(ctx, &req)
+	return eris.Wrap(err, "")
+}
+func (r *routerImpl) QueryTransactions(ctx context.Context, req *shardtypes.QueryTransactionsRequest) (
+	*shardtypes.QueryTransactionsResponse,
+	error,
+) {
+	res, err := r.ShardQuerier.Transactions(ctx, req)
+	return res, eris.Wrap(err, "")
+}
+
+func transactionToProto(sp *sign.Transaction) *shard.Transaction {
```

Comment:
would move this router/codec

Reply:
this is a bit too specific to `Router`s code to go in something like codec IMO. also this is just converting an object to another, so not really encoded related here.

Reply:
just rename it then

Reply:
this method isn't really a codec type of thing though. its just object conversion.

Reply:
i don't really mind what you call, goal here is to move it out of router.go

Reply:
i don't agree that its clutter though. its one function.

Reply:
ill just inline it instead.